src/dfx/src/lib/operations/canister/install_canister.rs

```diff
@@ -136,7 +140,47 @@ pub async fn install_canister(
         )
         .await?;
     }
-
+    let mut times = 0;
+    loop {
+        match agent
+            .read_state_canister_info(canister_id, "module_hash", false)
+            .await
+        {
+            Ok(reported_hash) => {
+                if reported_hash == new_hash {
+                    break;
+                } else if installed_module_hash
+                    .as_deref()
+                    .map_or(true, |old_hash| old_hash == reported_hash)
+                {
+                    times += 1;
+                    if times % 4 == 0 {
+                        info!(
+                            env.get_logger(),
+                            "Waiting for module change to be reflected in system state tree..."
+                        )
+                    }
+                    sleep(Duration::from_millis(500)).await;
+                } else {
+                    bail!("The reported module hash is neither the existing module or the new one. \
+                        It has likely been modified while this command is running. \
+                        The state of the canister is unknown. \
+                        For this reason, no post-installation tasks have been run, including asset uploads.")
+                }
+            }
+            Err(AgentError::LookupPathAbsent(_) | AgentError::LookupPathUnknown(_)) => {
+                times += 1;
+                if times % 4 == 0 {
+                    info!(
+                        env.get_logger(),
+                        "Waiting for module change to be reflected in system state tree..."
+                    )
+                }
+                sleep(Duration::from_millis(500)).await;
```

Comment:
What would you think about using a Waiter here, with some finite timeout?

Reply:
I'm not really sure how Waiter works, but it looks blocking, which doesn't play well with async.

Reply:
Wait, is that how we managed to starve the single-threaded web server however long ago?

Reply:
That's why we use Garcon, which provides async_await().  Example:

https://github.com/dfinity/sdk/blob/master/src/dfx/src/commands/ledger/mod.rs#L165

Reply:
Hmm. Some of those just block the thread anyway and then return a ready future, and others spin a new thread and sleep it. Both of these things defeat all benefits of asynchrony.